File: 0144_binaryTreePreorderTraversal/main.py
from typing import Optional, List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        def inorder(node):
            if not node:
                return
            res.append(node.val)
            inorder(node.left)
            inorder(node.right)

        inorder(root)
        
        return res
    
class Solution1:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        stack = []
        cur = root

        while cur or stack:
            while cur:
                stack.append(cur)
                res.append(cur.val)
                cur = cur.left
            cur = stack.pop()
            cur = cur.right
        
        return res
    
class Solution2:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        res = []
        stack = [root]

        while stack:
            cur = stack.pop()
            res.append(cur.val)
            if cur.right:
                stack.append(cur.right)
            if cur.left:
                stack.append(cur.left)
        return res        

File: 0144_binaryTreePreorderTraversal/main.cpp
#include <vector>
#include <stack>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    std::vector<int> preorderTraversal(TreeNode* root) {
        std::vector<int> res;
        std::stack<TreeNode*> st;

        while (root || !st.empty()) {
            while (root) {
                res.push_back(root -> val);
                st.push(root);
                root = root -> left;
            }
            root = st.top();
            st.pop();
            root = root -> right;
        }
        return res;
    }
};

class Solution1 {
public:
    std::vector<int> preorderTraversal(TreeNode* root) {
        std::vector<int> res;
        inorder(root, res);
        return res;
    }
private:
    void inorder(TreeNode* root, std::vector<int>& res) {
        if (!root) {
            return;
        }
        res.push_back(root -> val);
        inorder(root -> left, res);
        inorder(root -> right, res);
    }
};

class Solution2 {
public:
    std::vector<int> preorderTraversal(TreeNode* root) {
        if(!root)
            return {};

        std::vector<int> res;
        std::stack<TreeNode*> st;
        st.push(root);
        TreeNode *cur;

        while (!st.empty()) {
            cur = st.top();
            st.pop();
            res.push_back(cur->val);
            if(cur->right)
                st.push(cur->right);
            if(cur->left)
                st.push(cur->left);
        }
        return res;
    }
};